File: C++/Algorithms/Searching/Linear_Search.cpp
// Title : Linear Search Implementetion Using C++
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 18/07/2020

#include <iostream>
#include <stdlib.h>

using namespace std;

// Linear Search Algorithm
int ** linearSearch(int array[],int size, int searchValue){
	// Variable to store count of matching elements
	int match_count = 0;
	// Array to store 
	int match_array[size];

	// Linear Search Loop
	for(int i = 0; i < size; i++){
		if(array[i] == searchValue){
			match_array[match_count] = i;
			match_count += 1;
		}
	}

	// Allocating Memory Dynamically to result 2D array
	int ** result = (int**)malloc(2*sizeof(int));

	// Allocating Memory Dynamically for Index 0 of result 2d array
	result[0] = (int*)malloc(sizeof(int));

	// Assigning Match Count Value to Result 2D array at index[0][0]
	result[0][0] = match_count;

	// Allocating Memory Dynamically for Index 1 of result 2d array
	result[1] = (int*)malloc(match_count*sizeof(int));

	// Assigning Match Index Values to Result 2D array at index[0][i]
	for(int i = 0; i < match_count; i++){
		result[1][i] = match_array[i];
	}

	// Returning Result 2D Array
	return result;
}

// / Function to display/print array
void displayArray(int array[],int size){
	cout << "[ ";
	for(int i = 0; i < size; i++){
		cout << array[i] << ", ";
	}
	cout << "]\n";
}


int main(){
	// Variable for array size
	int size = 0;

	// Variable to store search value
	int searchValue = 0;

	// Array Size Input
	cout << "Enter Array Size : ";
	cin >> size;

	// Data Array
	int array[size];

	// Array Data Input
    for(int i = 0; i < size; i++){
    	cout << "Index [" << i << "] : ";
    	cin >> array[i];
    }


    // Search value input
    cout << "\nEnter Value to Search : ";
    cin >> searchValue;

    // Linear Search
    int ** result = linearSearch(array,size,searchValue);

    // Printing Match Result if match count is not 0
    if(result[0][0] != 0){
    	cout << "Matches Found : ";
    	displayArray(result[1],result[0][0]);
    }
    else{
    	cout << "No Matches Found!";
    }

	return 0;
}

File: JavaScript/Algorithms/Searching/Linear_Search.js
// Title : Linear Search Implementation Using JavaScript
// Author : Ansh Walia <anshwalia@outlook.com>
// Date Created : 18/07/2020

// Note : You can use Node.JS to run this code in terminal or,
// 	      You can directly run this code in browser console in developer tools

'use strict';

// Linear Search Algorithm
function linearSearch(array,value){
	// Array Object to store match found
	let matches = [];

	// Search Loop
	for(let i = 0; i < array.length; i++){
		if(array[i] == value){
			matches.push(i);
		}
	}

	// Returning Matches Found
	return matches;
}

// Function to test linear search algorithm
async function test(){
	// Test Array/Dataset
	const test_array = [1,2,3,4,5,6,2,2,5,7,8];
	
	// Test Value to Search
	const search_value = 2;

	// Calling Linear Search Function
	const result = await linearSearch(test_array,search_value);

	if(result.length != 0){
		console.log('Matches Found at Index : ',result);
	}
	else{
		console.log('No Matches Found!');
	}
}

// Entry Point
test();